File: matlab/tests/test_rsvd.m
function test_rsvd()
%test_rsvd  Correctness checks for randlapack.rsvd.
%
% Runs RSVD on matrices with a known rank structure and checks:
%   1. Orthogonality of U and V.
%   2. Reconstruction: A ~= U*S*V' when A is exactly low rank.
%   3. Near-optimality against the deterministic rank-k truncation
%      (Eckart-Young lower bound: no rank-k approximation beats it).
%   4. Output shapes and svd-compatible calling conventions.
%   5. Reproducibility from an explicit RNG state.
%   6. Error paths.
%
% Run from MATLAB after:
%   addpath('/path/to/randlapack-bindings/matlab')
%   test_rsvd

    types = {'double', 'single'};

    for ti = 1:numel(types)
        T = types{ti};

        % --- Exactly rank-r input: RSVD should reconstruct it ---
        m = 600; n = 300; r = 40;
        A = cast(randn(m, r) * randn(r, n), T);

        [U, S, V] = randlapack.rsvd(A, r);

        check_orthogonality(U, T, 'U');
        check_orthogonality(V, T, 'V');
        assert(size(U, 1) == m && size(V, 1) == n, 'wrong output row counts');
        assert(size(S, 1) == size(S, 2), 'S must be square');
        k_out = size(S, 1);
        assert(k_out <= r, 'returned rank %d exceeds requested %d', k_out, r);

        err = norm(double(A) - double(U) * double(S) * double(V)', 'fro') ...
              / norm(double(A), 'fro');
        tol = sqrt(double(m * n)) * eps(T) * 100;
        assert(err < tol, ...
               '%s: rank-%d reconstruction failed: rel err %g > tol %g', ...
               T, r, err, tol);

        fprintf('PASS: %s exact rank-%d reconstruction (rel err %.2e)\n', ...
                T, r, err);

        % --- Decaying spectrum: compare against the optimal truncation ---
        % Eckart-Young says the deterministic rank-k truncation is optimal,
        % so a randomized method can only match or lose. Requiring the loss
        % to stay within a modest factor is the real quality check; asking
        % for a fixed small error would only test the test matrix.
        k = 30;
        Ad = decaying_matrix(400, 200, T);
        [U, S, V] = randlapack.rsvd(Ad, k, 'p', 2);
        rand_err = norm(double(Ad) - double(U) * double(S) * double(V)', 'fro');

        [Us, Ss, Vs] = svd(double(Ad), 'econ');
        kk  = size(S, 1);
        opt = Us(:, 1:kk) * Ss(1:kk, 1:kk) * Vs(:, 1:kk)';
        opt_err = norm(double(Ad) - opt, 'fro');

        assert(rand_err <= 1.5 * opt_err + 10 * eps(T), ...
               ['%s: randomized rank-%d error %g is more than 1.5x the ' ...
                'optimal truncation error %g'], T, kk, rand_err, opt_err);
        fprintf('PASS: %s decaying spectrum, rank %d: %.4e vs optimal %.4e\n', ...
                T, kk, rand_err, opt_err);

        % --- Singular values: descending, positive, close to the true ones ---
        s = diag(S);
        assert(all(s > 0), 'singular values must be positive');
        assert(all(diff(double(s)) <= 1e-6 * double(s(1))), ...
               'singular values must be non-increasing');
        s_true = diag(Ss);
        rel_s  = abs(double(s(1)) - s_true(1)) / s_true(1);
        assert(rel_s < 0.05, ...
               '%s: leading singular value off by %.2f%%', T, 100 * rel_s);
        fprintf('PASS: %s leading singular value within %.2f%%\n', T, 100 * rel_s);
    end

    % --- svd-compatible calling conventions ---
    A = randn(300, 100);
    s_only = randlapack.rsvd(A, 10);
    assert(isvector(s_only) && numel(s_only) <= 10, ...
           'one-output form must return a vector of singular values');
    [U, S, ~] = randlapack.rsvd(A, 10);
    assert(ismatrix(S) && size(S, 1) == size(S, 2), ...
           'three-output form must return S as a square matrix');
    assert(norm(sort(double(s_only), 'descend') - sort(diag(double(S)), 'descend')) ...
           < 1e-8 * double(s_only(1)), ...
           'the two calling forms must agree on the singular values');
    fprintf('PASS: svd-compatible output conventions\n');

    % --- Reproducibility from an explicit state ---
    st = struct('counter', uint32([0 0 0 0]), 'key', uint32([0 7]));
    [U1, S1, ~] = randlapack.rsvd(A, 10, 'state', st);
    [U2, S2, ~] = randlapack.rsvd(A, 10, 'state', st);
    assert(isequal(U1, U2) && isequal(S1, S2), ...
           'same explicit state must give bit-identical results');
    fprintf('PASS: reproducible from an explicit RNG state\n');

    % --- The advanced state actually advances ---
    [~, ~, ~, st_out] = randlapack.rsvd(A, 10, 'state', st);
    assert(~isequal(st_out.counter, st.counter) || ~isequal(st_out.key, st.key), ...
           'returned state must differ from the input state');
    fprintf('PASS: RNG state advances\n');

    % --- Error paths ---
    check_error(@() randlapack.rsvd(zeros(0, 5), 1), 'randlapack:rsvd:emptyInput');
    check_error(@() randlapack.rsvd(A, 500), 'randlapack:rsvd:rankTooLarge');
    check_error(@() randlapack.rsvd(A, 0), 'MATLAB:rsvd:expectedPositive');
    check_error(@() randlapack.rsvd(A, 10, 'nosuchopt', 1), ...
                'randlapack:rsvd:unknownOption');
    check_error(@() randlapack.rsvd(A, 10, 'tol'), 'randlapack:rsvd:nargin');
    check_error(@() randlapack.rsvd(int32(A), 10), 'MATLAB:rsvd:invalidType');
    fprintf('PASS: error paths\n');

    fprintf('test_rsvd: all checks passed\n');
end


function A = decaying_matrix(m, n, T)
% Geometrically decaying spectrum with orthogonal factors, so the true
% singular values are known and the optimal truncation is computable.
    [Q1, ~] = qr(randn(m, min(m, n)), 0);
    [Q2, ~] = qr(randn(n, min(m, n)), 0);
    s = 0.85 .^ (0:min(m, n) - 1);
    A = cast(Q1 * diag(s) * Q2', T);
end


function check_orthogonality(X, T, name)
    [rows, k] = size(X);
    I   = cast(eye(k), T);
    err = norm(double(X)' * double(X) - double(I), 'fro');
    tol = sqrt(double(rows * k)) * eps(T) * 100;
    assert(err < tol, ...
           '%s orthogonality failed: ||%s^T %s - I||_F = %g > tol = %g', ...
           name, name, name, err, tol);
end


function check_error(fn, expected_id)
    try
        fn();
    catch err
        assert(strcmp(err.identifier, expected_id), ...
               'expected error %s, got %s', expected_id, err.identifier);
        return;
    end
    error('expected error %s, but no error was raised', expected_id);
end

File: python/tests/test_rsvd.py
"""Correctness checks for randlapack.rsvd."""
import numpy as np
import pytest

import randlapack as rl


DTYPES = [np.float64, np.float32]


def tol_for(dtype, m, n):
    return np.sqrt(float(m * n)) * np.finfo(dtype).eps * 100


def decaying_matrix(m, n, dtype, seed=0):
    """Geometric spectrum with orthogonal factors: the true singular values
    are known, so the optimal rank-k truncation is computable exactly."""
    rng = np.random.default_rng(seed)
    Q1, _ = np.linalg.qr(rng.standard_normal((m, min(m, n))))
    Q2, _ = np.linalg.qr(rng.standard_normal((n, min(m, n))))
    s = 0.85 ** np.arange(min(m, n))
    return np.asfortranarray((Q1 * s) @ Q2.T).astype(dtype)


@pytest.mark.parametrize("dtype", DTYPES)
def test_exact_low_rank_reconstruction(dtype):
    m, n, r = 600, 300, 40
    rng = np.random.default_rng(0)
    A = np.asfortranarray(
        (rng.standard_normal((m, r)) @ rng.standard_normal((r, n))).astype(dtype))

    U, s, V = rl.rsvd(A, r)

    assert U.shape[0] == m and V.shape[0] == n
    k = s.shape[0]
    assert k <= r
    assert U.shape[1] == k and V.shape[1] == k

    err = np.linalg.norm(A - U @ np.diag(s) @ V.T) / np.linalg.norm(A)
    assert err < tol_for(dtype, m, n), f"rel err {err:g}"


def orthogonality_error(X):
    k = X.shape[1]
    return np.linalg.norm(
        X.T.astype(np.float64) @ X.astype(np.float64) - np.eye(k))


@pytest.mark.parametrize("dtype", DTYPES)
def test_factors_are_orthonormal_well_conditioned(dtype):
    """On a well-conditioned input, both factors are orthonormal to
    machine precision."""
    m, n, r = 600, 300, 40
    rng = np.random.default_rng(1)
    A = np.asfortranarray(
        (rng.standard_normal((m, r)) @ rng.standard_normal((r, n))).astype(dtype))
    U, s, V = rl.rsvd(A, r)
    k = s.shape[0]
    assert orthogonality_error(U) < tol_for(dtype, m, k)
    assert orthogonality_error(V) < tol_for(dtype, n, k)


@pytest.mark.parametrize("dtype", DTYPES)
def test_orthogonality_follows_the_cholesky_qr_model(dtype):
    """U and V are orthonormal to DIFFERENT accuracies, by construction.

    V comes straight out of LAPACK's gesdd, so it is orthonormal to machine
    precision whatever the conditioning. U is formed as Q @ UT, and Q comes
    from the QB chain's CholeskyQR orthogonalizer, whose orthogonality loss
    scales like u * cond(A_sketch)**2 -- not u. On an ill-conditioned input
    U is therefore much looser than V, and power iterations do not fix it
    (they improve the approximation while making the sketch no better
    conditioned).

    Measured here (400x200, geometric spectrum, k=30): U ~ 3e-11 vs V ~ 7e-15
    in float64, U ~ 1.5e-2 vs V ~ 3e-6 in float32. This test asserts the
    model, so it catches a genuine regression without failing on expected
    numerical behaviour.
    """
    A = decaying_matrix(400, 200, dtype, seed=1)
    U, s, V = rl.rsvd(A, 30)
    k = s.shape[0]
    u = np.finfo(dtype).eps
    kappa = float(s[0] / s[-1])

    # V: machine precision.
    assert orthogonality_error(V) < tol_for(dtype, A.shape[1], k)

    # U: bounded by the CholeskyQR model, with a factor-100 safety margin
    # for the constants the model omits.
    bound = 100 * u * kappa ** 2
    err = orthogonality_error(U)
    assert err < bound, f"||U^T U - I|| = {err:g} exceeds u*kappa^2 bound {bound:g}"


@pytest.mark.parametrize("dtype", DTYPES)
def test_near_optimal_versus_truncated_svd(dtype):
    """Eckart-Young: the deterministic rank-k truncation is optimal, so the
    randomized method can only match it or lose. Bounding the loss is the
    real check; a fixed error threshold would only test the test matrix."""
    A = decaying_matrix(400, 200, dtype, seed=2)
    U, s, V = rl.rsvd(A, 30, p=2)
    k = s.shape[0]

    rand_err = np.linalg.norm(
        A.astype(np.float64) - (U @ np.diag(s) @ V.T).astype(np.float64))

    Us, ss, Vh = np.linalg.svd(A.astype(np.float64), full_matrices=False)
    opt = (Us[:, :k] * ss[:k]) @ Vh[:k, :]
    opt_err = np.linalg.norm(A.astype(np.float64) - opt)

    assert rand_err <= 1.5 * opt_err + 10 * np.finfo(dtype).eps, (
        f"randomized {rand_err:g} vs optimal {opt_err:g}")


@pytest.mark.parametrize("dtype", DTYPES)
def test_singular_values_descending_and_accurate(dtype):
    A = decaying_matrix(400, 200, dtype, seed=3)
    _, s, _ = rl.rsvd(A, 30, p=2)
    assert np.all(s > 0)
    assert np.all(np.diff(s.astype(np.float64)) <= 1e-6 * float(s[0]))

    s_true = np.linalg.svd(A.astype(np.float64), compute_uv=False)
    assert abs(float(s[0]) - s_true[0]) / s_true[0] < 0.05


def test_v_is_untransposed():
    """rsvd returns V, not vh. Verify against numpy's convention directly."""
    A = decaying_matrix(300, 150, np.float64, seed=4)
    U, s, V = rl.rsvd(A, 20, p=2)
    assert V.shape == (A.shape[1], s.shape[0])
    # Reconstruction works with V.T, and would be a shape error with V.
    approx = U @ np.diag(s) @ V.T
    assert approx.shape == A.shape


def test_c_order_input_matches_fortran_order():
    A_c = np.random.default_rng(5).standard_normal((600, 200))
    A_f = np.asfortranarray(A_c)
    assert not A_c.flags.f_contiguous

    U1, s1, V1 = rl.rsvd(A_c, 20, state=5)
    U2, s2, V2 = rl.rsvd(A_f, 20, state=5)

    assert np.array_equal(s1, s2)
    assert np.array_equal(U1, U2)
    assert np.array_equal(V1, V2)


def test_reproducible_from_explicit_state():
    A = np.asfortranarray(
        np.random.default_rng(6).standard_normal((500, 200)))
    st = {"counter": np.zeros(4, dtype=np.uint32),
          "key": np.array([0, 7], dtype=np.uint32)}
    U1, s1, V1 = rl.rsvd(A, 20, state=st)
    U2, s2, V2 = rl.rsvd(A, 20, state=st)
    assert np.array_equal(U1, U2)
    assert np.array_equal(s1, s2)
    assert np.array_equal(V1, V2)


def test_state_advances():
    A = np.asfortranarray(
        np.random.default_rng(7).standard_normal((500, 200)))
    st = {"counter": np.zeros(4, dtype=np.uint32),
          "key": np.array([0, 7], dtype=np.uint32)}
    _, _, _, st_out = rl.rsvd(A, 20, state=st, return_state=True)
    assert not (np.array_equal(st_out["counter"], st["counter"])
                and np.array_equal(st_out["key"], st["key"]))


def test_tolerance_can_truncate_below_the_requested_rank():
    """A loose tol lets QB stop early, so k_out < k. The returned shapes are
    the only place that rank is reported."""
    A = decaying_matrix(400, 200, np.float64, seed=8)
    _, s_tight, _ = rl.rsvd(A, 60, tol=0.0, block_sz=4)
    _, s_loose, _ = rl.rsvd(A, 60, tol=1e-2, block_sz=4)
    assert s_loose.shape[0] <= s_tight.shape[0]


def test_error_paths():
    A = np.random.default_rng(9).standard_normal((300, 100))
    with pytest.raises(ValueError):
        rl.rsvd(np.zeros((0, 5)), 1)
    with pytest.raises(ValueError):
        rl.rsvd(A, 500)             # k > min(m, n)
    with pytest.raises(ValueError):
        rl.rsvd(A, 0)
    with pytest.raises(ValueError):
        rl.rsvd(A, 10, tol=-1.0)
    with pytest.raises(ValueError):
        rl.rsvd(A, 10, block_sz=0)
    with pytest.raises(ValueError):
        rl.rsvd(A, 10, p=-1)
    with pytest.raises(TypeError):
        rl.rsvd(A.astype(np.int32), 10)
